app/controllers/dynamic_pages_controller.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+class DynamicPagesController < ApplicationController
+  before_action :require_admin, except: [:index, :show]
+  def index
+    if logged_in?
+      @current_user_is_admin = is_admin?
+      @current_user_is_teacher = is_teacher?
+      @current_user_is_verified_teacher = is_verified_teacher?
+    end
+    @all_dynamic_pages = DynamicPage.all
+  end
+  def delete
+    DynamicPage.destroy(params[:id])
+    redirect_to dynamic_pages_path
+  end
+  def new
+    @dynamic_page = DynamicPage.new(flash[:dynamic_page])
+  end
+  def create
+    @dynamic_page = DynamicPage.new(dynamic_page_params)
+    if DynamicPage.find_by(slug: @dynamic_page.slug)
+      flash[:dynamic_page] = params[:dynamic_page]
+      redirect_to({ action: "new" }, alert:  "That slug already exists :(")
+    elsif @dynamic_page.save
+      flash[:success] = "Created #{@dynamic_page.title} page successfully."
+      redirect_to ({ action: "show", slug: @dynamic_page.slug })
+    else
+      redirect_to root_path, alert: "Failed to submit information :("
+    end
+  end
+  def show
+    @dynamic_page = DynamicPage.find_by(slug: params[:slug])
+    if @dynamic_page.permissions == "Admin" && !is_admin?
+      redirect_to dynamic_pages_path, alert: "You do not have permission to view that page!"
+    elsif @dynamic_page.permissions == "Verified Teacher" && (!is_admin?) && (!is_verified_teacher?)
+      redirect_to dynamic_pages_path, alert: "You do not have permission to view that page!"
+    end
+  end
+  def edit
+    @dynamic_page = DynamicPage.find_by(slug: params[:slug])
+    @dynamic_page.assign_attributes(flash[:dynamic_page])
+  end
+  def update
```

Comment:
This method could use a bit of cleanup. 

* If we move the uniqueness constraints on the `slug` to the DB then we can simplify a lot of code. It's not perfect, but we can pass the ActiveRecord error message to the frontend if there is a failure.
* I think we should explicitly set `last_editor = current_user.id` in here, to be clear what's going on.

Reply:
As far as I can tell, we still need to do a check to make sure the slug isn't already in the database because @dynamic_page.save throws ActiveRecord::RecordNotUnique exception if @dynamic_page has a duplicate slug.

Reply:
![image](https://user-images.githubusercontent.com/71572838/165878256-b0a291cd-4fe1-42f8-8f6e-7adacfd0eb36.png)

Reply:
Huh! Thanks for trying this out. I could have sworn `@thing.save` isn't supposed to raise an error (since that's why we have `save!`. But I think what's also going on is that we don't have a _model_ validation on the `slug` column. We should be able to validate `uniqueness: true` in our model then Rails will at least take care of this logic for us. :) ActiveRecord errors are definitely caught by `save`. Though you can call call `.valid?` if you wish.